src/Agda2Hs/Compile/Type.hs

```diff
@@ -210,16 +209,30 @@ compileInlineType f args = do
     YesReduction _ t -> compileType t
     _                -> genericDocError =<< text "Could not reduce inline type alias " <+> prettyTCM f
 
+
+-- Compile a function domain.
+-- A domain can either be:
+--
+-- - dropped if the argument is erased.
+-- - added as a class constraint.
+-- - kept as a regular explict argument.
 compileDom :: ArgName -> Dom Type -> C CompiledDom
 compileDom x a
   | usableModality a = case getHiding a of
       Instance{} -> DomConstraint . Hs.TypeA () <$> compileType (unEl $ unDom a)
-      NotHidden  -> uncurry DomType <$> compileType' (unEl $ unDom a)
-      Hidden     ->
-        ifM (canErase $ unDom a)
-            (return DomDropped)
-            (genericDocError =<< do text "Implicit type argument not supported: " <+> prettyTCM x)
-  | otherwise    = return DomDropped
+      _          -> ifM (canErase $ unDom a) (return DomDropped)
+                        (uncurry DomType <$> compileTypeStrictness (unEl $ unDom a))
+  | otherwise = return DomDropped
+
+data DomOutput = DOInstance | DODropped | DOKept
+
+compiledDom :: Dom Type -> C DomOutput
```

Comment:
Why is this a separate function from `compileDom`? Seems like it's a potential source of subtle bugs if these two don't match.

Reply:
The reason was `compileDom` atually compiles all the types to Haskell, so if this was the one being used in the typed-compilation we would be constantly compiling types to Haskell, even though we only care about whether those things are kept or dropped.
If we had dependent types it would be easier to express at the type-level that `compileDom` is now a refinement of `compiledDom`, but I'll just implement the former with the latter.